addons/web/controllers/main.py

```diff
@@ -990,140 +990,155 @@ def action(self, model, id):
-                content_type = res['file_type']
-        else:
-            res = Model.default_get(cr, uid, fields, context)
-        filecontent = base64.b64decode(res.get(field) or '')
-        if not filecontent:
-            return request.not_found()
-        else:
-            filename = '%s_%s' % (model.replace('.', '_'), id)
-            if filename_field:
-                filename = res.get(filename_field, '') or filename
-            return request.make_response(filecontent,
-                [('Content-Type', content_type),
-                 ('Content-Disposition', content_disposition(filename))])
-
-    @http.route('/web/binary/saveas_ajax', type='http', auth="public")
-    @serialize_exception
-    def saveas_ajax(self, data, token):
-        jdata = simplejson.loads(data)
-        model = jdata['model']
-        field = jdata['field']
-        data = jdata['data']
-        id = jdata.get('id', None)
-        filename_field = jdata.get('filename_field', None)
-        context = jdata.get('context', {})
-        content_type = 'application/octet-stream'
+    # return
+    return (status, headers, content)
 
-        Model = request.session.model(model)
-        fields = [field]
-        if filename_field:
-            fields.append(filename_field)
+class Binary(http.Controller):
+
+    @http.route(['/web/content',
+        '/web/content/<string:xmlid>',
+        '/web/content/<string:xmlid>/<string:filename>',
+        '/web/content/<int:id>',
+        '/web/content/<string:model>/<int:id>/<string:field>',
+        '/web/content/<string:model>/<int:id>/<string:field>/<string:filename>'], type='http', auth="public")
+    def content_common(self, xmlid=None, model='ir.attachment', id=None, field='datas', filename_field='datas_fname', unique=None, filename=None, mimetype=None, extension=None, download=None, data=None, token=None):
+        status, headers, content = binary_content(xmlid=xmlid, model=model, id=id, field=field, unique=unique, filename=filename, filename_field=filename_field, download=download, mimetype=mimetype, default_mimetype='image/png')
         if data:
-            res = {field: data, filename_field: jdata.get('filename', None)}
-        elif id:
-            fields.append('file_type')
-            res = Model.read([int(id)], fields, context)[0]
-            if res.get('file_type'):
-                content_type = res['file_type']
-        else:
-            res = Model.default_get(fields, context)
-        filecontent = base64.b64decode(res.get(field) or '')
-        if not filecontent:
-            raise ValueError(_("No content found for field '%s' on '%s:%s'") %
-                (field, model, id))
+            content = base64.b64decode(data)
+        elif status == 304:
+            return werkzeug.wrappers.Response(status=304, headers=headers)
+        elif status == 404:
+            return request.not_found()
+
+        content_base64 = content and base64.b64decode(content) or self.placeholder()
```

Comment:
That's more subtle.
Attachments are directly store in base64 in database, but normal Binary fields are not [0].
Thought, when reading data from database, it may be encoded if it contains low bytes values [1]
You should test with a text file save in this field [2]
[0] https://github.com/odoo/odoo/blob/00fa97926485ccfcefc07e5388d886dd15c39bb5/openerp/fields.py#L1363-L1364
[1] https://github.com/odoo/odoo/blob/00fa97926485ccfcefc07e5388d886dd15c39bb5/openerp/osv/fields.py#L1090-L1099
[2] https://github.com/odoo/odoo/blob/4532ca61966be8ca135601fa010953d11fbb640b/openerp/addons/test_converter/models.py#L13